dm/master/server_test.go

```diff
@@ -428,6 +441,180 @@ func (t *testMaster) TestStartTask(c *check.C) {
+		c.Errorf("db unfulfilled expectations: %s", err)
+	}
+	ifm, _, err := pessimism.GetAllInfo(etcdTestCli)
+	c.Assert(err, check.IsNil)
+	c.Assert(ifm, check.HasLen, 0)
+	opm, _, err := pessimism.GetAllOperations(etcdTestCli)
+	c.Assert(err, check.IsNil)
+	c.Assert(opm, check.HasLen, 0)
+	clearSchedulerEnv(c, cancel, &wg)
+
+	// test remove meta with optimist
+	ctx, cancel = context.WithCancel(context.Background())
+	cfg.ShardMode = config.ShardOptimistic
+	req = &pb.StartTaskRequest{
+		Task:       strings.ReplaceAll(taskConfig, `shard-mode: ""`, fmt.Sprintf(`shard-mode: "%s"`, cfg.ShardMode)),
+		Sources:    sources,
+		RemoveMeta: true,
+	}
+	server.scheduler, _ = testMockScheduler(ctx, &wg, c, sources, workers, "",
+		makeWorkerClientsForHandle(ctrl, taskName, sources, workers, req))
+	server.pessimist = shardddl.NewPessimist(&logger, func(task string) []string { return sources })
+	server.optimist = shardddl.NewOptimist(&logger)
+
+	var (
+		p           = parser.New()
+		se          = tidbmock.NewContext()
+		tblID int64 = 111
+
+		st1      = optimism.NewSourceTables(taskName, sources[0])
+		DDLs1    = []string{"ALTER TABLE bar ADD COLUMN c1 INT"}
+		tiBefore = createTableInfo(c, p, se, tblID, `CREATE TABLE bar (id INT PRIMARY KEY)`)
+		tiAfter1 = createTableInfo(c, p, se, tblID, `CREATE TABLE bar (id INT PRIMARY KEY, c1 TEXT)`)
+		info1    = optimism.NewInfo(taskName, sources[0], "foo-1", "bar-1", schema, table, DDLs1, tiBefore, tiAfter1)
+		op1      = optimism.NewOperation(ID, taskName, sources[0], info1.UpSchema, info1.UpTable, DDLs1, optimism.ConflictNone, false)
+	)
+
+	_, err = optimism.PutSourceTables(etcdTestCli, st1)
+	c.Assert(err, check.IsNil)
+	_, err = optimism.PutInfo(etcdTestCli, info1)
+	c.Assert(err, check.IsNil)
+	_, succ, err = optimism.PutOperation(etcdTestCli, false, op1)
+	c.Assert(succ, check.IsTrue)
+	c.Assert(err, check.IsNil)
+
+	err = server.pessimist.Start(ctx, etcdTestCli)
+	c.Assert(err, check.IsNil)
+	err = server.optimist.Start(ctx, etcdTestCli)
+	c.Assert(err, check.IsNil)
+
+	db, mock, err = sqlmock.New()
+	c.Assert(err, check.IsNil)
+	conn.DefaultDBProvider = &mockDBProvider{db: db}
+	mock.ExpectBegin()
+	mock.ExpectExec(fmt.Sprintf("DROP TABLE IF EXISTS `%s`.`%s`", cfg.MetaSchema, cputil.LoaderCheckpoint(cfg.Name))).WillReturnResult(sqlmock.NewResult(1, 1))
+	mock.ExpectExec(fmt.Sprintf("DROP TABLE IF EXISTS `%s`.`%s`", cfg.MetaSchema, cputil.SyncerCheckpoint(cfg.Name))).WillReturnResult(sqlmock.NewResult(1, 1))
+	mock.ExpectExec(fmt.Sprintf("DROP TABLE IF EXISTS `%s`.`%s`", cfg.MetaSchema, cputil.SyncerShardMeta(cfg.Name))).WillReturnResult(sqlmock.NewResult(1, 1))
+	mock.ExpectExec(fmt.Sprintf("DROP TABLE IF EXISTS `%s`.`%s`", cfg.MetaSchema, cputil.SyncerOnlineDDL(cfg.Name))).WillReturnResult(sqlmock.NewResult(1, 1))
+	mock.ExpectCommit()
+	c.Assert(len(server.optimist.Locks()), check.Greater, 0)
+
+	resp, err = server.StartTask(context.Background(), req)
```

Comment:
`StartTask` with same name again when the previous call is still running (in L513)?